gcc/rust/backend/rust-compile-intrinsic.cc

```diff
@@ -500,6 +504,34 @@ transmute_intrinsic_handler (Context *ctx, TyTy::BaseType *fntype_tyty)
   tree convert_me_expr
     = ctx->get_backend ()->var_expression (convert_me_param, Location ());
 
+  // check for transmute pre-conditions
+  tree target_type_expr = TREE_TYPE (DECL_RESULT (fndecl));
+  tree source_type_expr = compiled_types.at (0);
+  tree target_size_expr = TYPE_SIZE (target_type_expr);
+  tree source_size_expr = TYPE_SIZE (source_type_expr);
+  // for some reason, unit types and other zero-sized types return NULL for the
+  // size expressions
+  unsigned HOST_WIDE_INT target_size
+    = target_size_expr ? TREE_INT_CST_LOW (target_size_expr) : 0;
+  unsigned HOST_WIDE_INT source_size
+    = source_size_expr ? TREE_INT_CST_LOW (source_size_expr) : 0;
+
+  // size check for concrete types
+  // TODO(liushuyu): check alignment for pointers; check for dependently-sized
+  // types
+  if (target_size != source_size)
+    {
+      rust_error_at (fntype->get_locus (),
```

Comment:
I am not sure how to get the original call-site location from the function declaration here.

Reply:
Ah yes good point there is no mechanism for this here. Let me take a look to see what might be the best approach for this and I will raise an issue with a guide to add this for someone to implement.

Reply:
Lets raise this as a seperate issue to get the call site info as this PR looks good to go now